moveit_ros/hybrid_planning/local_planner/local_constraint_solver_plugins/src/servo_solver.cpp

```diff
@@ -0,0 +1,200 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr, Adam Pettinger
+ */
+
+#include <control_msgs/msg/joint_jog.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <moveit/local_constraint_solver_plugins/servo_solver.hpp>
+#include <moveit/planning_scene/planning_scene.h>
+#include <moveit/robot_state/conversions.h>
+#include <moveit/utils/logger.hpp>
+#include <local_planner_parameters.hpp>
+
+namespace moveit::hybrid_planning
+{
+namespace
+{
+rclcpp::Logger getLogger()
+{
+  return moveit::getLogger("servo_solver");
+}
+}  // namespace
+
+bool ServoSolver::initialize(const rclcpp::Node::SharedPtr& node,
+                             const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
+                             const std::string& /*group_name*/)
+{
+  planning_scene_monitor_ = planning_scene_monitor;
+  node_ = node;
+
+  const std::shared_ptr<const servo_solver_parameters::ParamListener> solver_param_listener =
+      std::make_shared<const servo_solver_parameters::ParamListener>(node, "");
+  solver_parameters_ = solver_param_listener->get_params();
+
+  // Get Servo Parameters
+  const std::string param_namespace = "moveit_servo";
+  const std::shared_ptr<const servo::ParamListener> servo_param_listener =
+      std::make_shared<const servo::ParamListener>(node, param_namespace);
+  servo_parameters_ = servo_param_listener->get_params();
+
+  // Create Servo and start it
+  servo_ = std::make_unique<moveit_servo::Servo>(node_, servo_param_listener, planning_scene_monitor_);
+
+  // Set servo publish_period
+  auto local_param_listener = local_planner_parameters::ParamListener(node_, "");
+  const auto local_config = local_param_listener.get_params();
+
+  // Publish period is the time difference used for numerical integration. It is recommended that the period 3x smaller
+  // than the actual publish period.
+  const double servo_publish_rate = 1.0 / 3.0 * local_config.local_planning_frequency;
+  if (!node_->has_parameter("publish_period"))
+  {
+    RCLCPP_ERROR(getLogger(), "Node used by servo solver doesn't seem to have a parameter called 'publish_period' that "
+                              "shouldn't happen!");
+    return false;
+  }
+  node_->set_parameter(rclcpp::Parameter("publish_period", servo_publish_rate));
+  return true;
+}
+
+bool ServoSolver::reset()
+{
+  RCLCPP_INFO(getLogger(), "Reset Servo Solver");
+  joint_cmd_rolling_window_.clear();
```

Comment:
Does this get called on every iteration of of local planning or is it once per global plan?

Reply:
Only once per global plan. Would it make sense to reset the rolling window when the reference pose changes?

Reply:
That is interesting. I thought it may have been called on each iteration. In that case, won't the `joint_cmd_rolling_window_` queue grow indefinitely as `solve` continues to be called? In that case, what I would do is remove old commands from `joint_cmd_rolling_window_` in the `solve` method. Then I think you can set the timestep of servo queal to the time step of the local planner.